File: Assets/Bundles/Shared/Resources/TS/Item/HeldItems/Damagers/MeleeHeldItem.lua
-- Compiled with unity-ts v2.1.0-75
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local EffectsManager = require("Shared/TS/Effects/EffectsManager").EffectsManager
local Entity = require("Shared/TS/Entity/Entity").Entity
local Bundle_ItemSword = require("Shared/TS/Util/ReferenceManagerResources").Bundle_ItemSword
local HeldItem = require("Shared/TS/Item/HeldItems/HeldItem").HeldItem
local MeleeHeldItem
do
	local super = HeldItem
	MeleeHeldItem = setmetatable({}, {
		__tostring = function()
			return "MeleeHeldItem"
		end,
		__index = super,
	})
	MeleeHeldItem.__index = MeleeHeldItem
	function MeleeHeldItem.new(...)
		local self = setmetatable({}, MeleeHeldItem)
		return self:constructor(...) or self
	end
	function MeleeHeldItem:constructor(...)
		super.constructor(self, ...)
	end
	function MeleeHeldItem:OnUseClient(useIndex)
		super.OnUseClient(self, useIndex)
		local meleeData = self.meta.melee
		if not meleeData then
			return nil
		end
		self:Log("Using Server")
		-- Only local player should do collisions checks
		-- TODO make sure other players show the attacks effects just without having to do collision checks
		if self.entity:IsLocalCharacter() then
			local entityDriver = self.entity:GetEntityDriver()
			entityDriver:UpdateSyncTick()
			local hitTargets = self:GetCollisions(meleeData.colliderData)
			local _hitTargets = hitTargets
			local _arg0 = function(data)
				local _condition = self.bundles
				if _condition then
					local _result = self.meta.melee
					if _result ~= nil then
						_result = _result.onHitPrefabId
					end
					_condition = _result
				end
				if _condition ~= 0 and (_condition == _condition and _condition) then
					-- Local damage predictions
					local effectGO = EffectsManager:SpawnBundleGroupEffect(self.bundles, Bundle_ItemSword.Prefabs, self.meta.melee.onHitPrefabId, data.hitPosition, Quaternion.LookRotation(data.hitDirection).eulerAngles)
					if effectGO then
						effectGO.transform.parent = data.hitEntity.model.transform
					end
				end
			end
			for _k, _v in _hitTargets do
				_arg0(_v, _k - 1, _hitTargets)
			end
		end
	end
	function MeleeHeldItem:OnUseServer(useIndex)
		print("melee.onUseServer")
		Profiler:BeginSample("Melee.OnUseServer")
		Profiler:BeginSample("super.OnUseServer")
		super.OnUseServer(self, useIndex)
		Profiler:EndSample()
		local meleeData = self.meta.melee
		if not meleeData then
			Profiler:EndSample()
			return nil
		end
		self:Log("Using Server")
		Profiler:BeginSample("GetCollisions")
		local hitTargets = self:GetCollisions(meleeData.colliderData)
		Profiler:EndSample()
		print("Server hit tick=" .. tostring(InstanceFinder.TimeManager.Tick) .. ", hitTargets=" .. tostring(#hitTargets))
		Profiler:BeginSample("HitTargetsInflictDamage")
		local _hitTargets = hitTargets
		local _arg0 = function(data)
			local _fn = (Flamework.resolveDependency("Bundles/Server/Services/Global/Damage/DamageService@DamageService"))
			local _exp = data.hitEntity
			local _result = meleeData
			if _result ~= nil then
				_result = _result.damage
			end
			local _condition = _result
			if _condition == nil then
				_condition = 0
			end
			local _object = {}
			local _left = "damageType"
			local _result_1 = meleeData
			if _result_1 ~= nil then
				_result_1 = _result_1.damageType
			end
			_object[_left] = _result_1
			_object.fromEntity = self.entity
			_object.knockbackDirection = data.knockbackDirection
			_fn:InflictDamage(_exp, _condition, _object)
		end
		for _k, _v in _hitTargets do
			_arg0(_v, _k - 1, _hitTargets)
		end
		Profiler:EndSample()
		Profiler:EndSample()
	end
	function MeleeHeldItem:GetCollisions(boxData)
		local collisionData = {}
		local closestCollisionData
		if not boxData then
			error(self.meta.displayName .. " Melee No Box Data")
			return collisionData
		end
		self:Log("Finding Collisions")
		local detectHalfSize = Vector3.new(boxData.boxHalfWidth, boxData.boxHalfHeight, boxData.boxHalfDepth)
		local layerMask = 8
		local _condition = boxData.localPositionOffsetX
		if _condition == nil then
			_condition = 0
		end
		local _exp = boxData.boxHalfHeight
		local _condition_1 = boxData.localPositionOffsetY
		if _condition_1 == nil then
			_condition_1 = 0
		end
		local _exp_1 = _exp + _condition_1
		local _exp_2 = -0.5 + boxData.boxHalfDepth
		local _condition_2 = boxData.localPositionOffsetZ
		if _condition_2 == nil then
			_condition_2 = 0
		end
		local boxLocalPos = Vector3.new(_condition, _exp_1, _exp_2 + _condition_2)
		local colliderWorldPos = self.entity.model.transform:TransformPoint(boxLocalPos)
		local hitColliders = Physics:OverlapBox(colliderWorldPos, detectHalfSize, self.entity.model.transform.rotation, layerMask, 0)
		local foundRaycastCollision
		local rayDistance = (boxData.boxHalfDepth + boxData.boxHalfHeight + boxData.boxHalfWidth) * 2
		-- For each collider in the box detection
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < hitColliders.Length) then
					break
				end
				self:Log("Collider: " .. tostring(i))
				local collider = hitColliders:GetValue(i)
				local targetEntity = Entity:FindByCollider(collider)
				-- If we hit an entity that is not the owner of this item
				if not targetEntity then
					-- Box check doesn't care about non entities
					continue
				end
				if targetEntity.id == self.entity.id then
					-- Hit Self
					continue
				end
				self:Log("Hit Entity: " .. tostring(targetEntity.id))
				-- Raycast to the target to find a more concrete collisions
				-- TODO the entities look direction should be synced here not just its plane aligned look direction
				local backtrackOffset = self.entity.model.transform:TransformVector(Vector3.new(0, 0, -0.5))
				local rayStart = self.entity:GetHeadPosition() + backtrackOffset
				local rayEnd = targetEntity:GetHeadPosition()
				local _rayEnd = rayEnd
				local _rayStart = rayStart
				local hitDirection = (_rayEnd - _rayStart).normalized
				-- RAYCAST ALL
				self:Log("Raycast All")
				local hitInfosArray = Physics:RaycastAll(rayStart, hitDirection, rayDistance, -1)
				local hitInfos = hitInfosArray
				local blockerDistance = 9999
				-- DebugUtil.DrawSingleLine(rayStart, rayEnd, Color.cyan, 2);
				-- Check each ray collision
				do
					local i = 0
					local _shouldIncrement_1 = false
					while true do
						if _shouldIncrement_1 then
							i += 1
						else
							_shouldIncrement_1 = true
						end
						if not (i < hitInfos.Length) then
							break
						end
						self:Log("Raycasting to target")
						local hitInfo = hitInfos:GetValue(i)
						-- Look for entities and blocking colliders
						local rayTarget = Entity:FindByCollider(hitInfo.collider)
						if rayTarget then
							if rayTarget.id == self.entity.id then
								-- Hit self, skip
								continue
							elseif rayTarget.id == targetEntity.id then
								self:Log("Raycast hit: " .. tostring(rayTarget.id))
								-- Raycast hit the target entity
								foundRaycastCollision = {
									hitEntity = targetEntity,
									hitDirection = hitDirection,
									hitPosition = hitInfo.point,
									hitNormal = hitInfo.normal,
									distance = hitInfo.distance,
									knockbackDirection = self.entity.gameObject.transform.forward,
								}
								-- DebugUtil.DrawSingleLine(rayStart, hitInfo.point, Color.green, 2);
								if not closestCollisionData or hitInfo.distance < closestCollisionData.distance then
									self:Log("New closest target: " .. tostring(hitInfo.distance))
									closestCollisionData = foundRaycastCollision
									continue
								end
							else
								-- Hit a non entity object
								self:Log("Blocked by object: " .. hitInfo.collider.gameObject.name)
								blockerDistance = math.min(blockerDistance, hitInfo.distance)
								-- DebugUtil.DrawSingleLine(rayStart, hitInfo.point, Color.red, 2);
							end
						end
						if foundRaycastCollision then
							self:Log("Found target")
							if foundRaycastCollision.distance < blockerDistance then
								self:Log("Using target")
								-- Not blocked by something
								local _collisionData = collisionData
								local _foundRaycastCollision = foundRaycastCollision
								table.insert(_collisionData, _foundRaycastCollision)
								-- DebugUtil.DrawSingleLine(rayStart, foundRaycastCollision.hitPosition, Color.green, 2);
							else
								self:Log("Can't use target because of blocker")
							end
						end
					end
				end
			end
		end
		-- DebugUtil.DrawBox(
		-- colliderWorldPos,
		-- this.entity.model.transform.rotation,
		-- detectHalfSize,
		-- closestCollisionData ? Color.green : Color.red,
		-- 2,
		-- );
		local _result = self.meta.melee
		if _result ~= nil then
			_result = _result.canHitMultipleTargets
		end
		if _result then
			return collisionData
		elseif closestCollisionData then
			return { closestCollisionData }
		end
		return {}
	end
end
return {
	MeleeHeldItem = MeleeHeldItem,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Item/HeldItems/Damagers/MeleeHeldItem.ts
import { Dependency } from "@easy-games/flamework-core";
import { Theme } from "Shared/Util/Theme";
import { DamageService } from "../../../../Server/Services/Global/Damage/DamageService";
import { EffectsManager } from "../../../Effects/EffectsManager";
import { Entity } from "../../../Entity/Entity";
import { Bundle_ItemSword } from "../../../Util/ReferenceManagerResources";
import { HeldItem } from "../HeldItem";

export class MeleeHeldItem extends HeldItem {
	private gizmoEnabled = true;
	private combatVars = DynamicVariablesManager.Instance.GetVars("Combat")!;

	override OnUseClient(useIndex: number) {
		super.OnUseClient(useIndex);
		let meleeData = this.meta.melee;
		if (!meleeData) {
			return;
		}
		this.Log("Using Server");
		//Only local player should do collisions checks
		//TODO make sure other players show the attacks effects just without having to do collision checks
		if (this.entity.IsLocalCharacter()) {
			const entityDriver = this.entity.GetEntityDriver();
			entityDriver.UpdateSyncTick();

			let farTargets = this.ScanForHits();

			for (const data of farTargets) {
				if (this.bundles && this.meta.melee?.onHitPrefabId) {
					//Local damage predictions
					const effectGO = EffectsManager.SpawnBundleGroupEffect(
						this.bundles,
						Bundle_ItemSword.Prefabs,
						this.meta.melee.onHitPrefabId,
						data.hitPosition,
						Quaternion.LookRotation(data.hitDirection).eulerAngles,
					);
					if (effectGO) {
						effectGO.transform.parent = data.hitEntity.model.transform;
					}
				}
			}
		}
	}

	override OnUseServer(useIndex: number) {
		print("melee.onUseServer");
		Profiler.BeginSample("Melee.OnUseServer");

		Profiler.BeginSample("super.OnUseServer");
		super.OnUseServer(useIndex);
		Profiler.EndSample();
		let meleeData = this.meta.melee;
		if (!meleeData) {
			Profiler.EndSample();
			return;
		}
		this.Log("Using Server");
		Profiler.BeginSample("GetCollisions");
		let hitTargets = this.ScanForHits();
		Profiler.EndSample();
		print("Server hit tick=" + InstanceFinder.TimeManager.Tick + ", hitTargets=" + hitTargets.size());
		Profiler.BeginSample("HitTargetsInflictDamage");
		hitTargets.forEach((data) => {
			Dependency<DamageService>().InflictDamage(data.hitEntity, meleeData?.damage ?? 0, {
				damageType: meleeData?.damageType,
				fromEntity: this.entity,
				knockbackDirection: data.knockbackDirection,
			});
		});
		Profiler.EndSample();
		Profiler.EndSample();
	}

	private ScanForHits(): MeleeHit[] {
		let farBox = this.combatVars.GetVector3("swordBoxFar");
		let closeBox = this.combatVars.GetVector3("swordBoxClose");

		let farHits = this.ScanBox(farBox, [], Theme.Red);
		let closeHits = this.ScanBox(
			closeBox,
			farHits.map((x) => x.hitEntity.id),
			Theme.Green,
		);
		let results = [...farHits, ...closeHits];
		return results;
	}

	private ScanBox(box: Vector3, ignoreEntityIds: number[], debugColor: Color): MeleeHit[] {
		let collisionData: Array<MeleeHit> = [];
		let closestCollisionData: MeleeHit | undefined;

		const layerMask = 8; // character layer: 1 << 3

		const lookVec = this.entity.entityDriver.GetLookVector();
		box = box.add(new Vector3(0, 0, 0.5));
		let halfExtents = new Vector3(box.x / 2, box.y / 2, box.z / 2);
		let headOffset = this.entity.GetHeadOffset();
		const t = this.entity.model.transform;
		let colliderWorldPos = t.position.add(headOffset).add(lookVec.mul(-0.5 + box.z / 2));

		let rotation = Quaternion.LookRotation(lookVec);

		if (this.gizmoEnabled) {
			DebugUtil.DrawBox(colliderWorldPos, rotation, halfExtents, debugColor, 2);
		}
		const hitColliders = Physics.OverlapBox(
			colliderWorldPos,
			halfExtents,
			rotation,
			layerMask,
			QueryTriggerInteraction.UseGlobal,
		);

		let foundRaycastCollision: MeleeHit | undefined;
		const rayDistance = box.magnitude;

		//For each collider in the box detection
		for (let i = 0; i < hitColliders.Length; i++) {
			this.Log("Collider: " + i);
			const collider = hitColliders.GetValue(i);
			const targetEntity = Entity.FindByCollider(collider);
			//If we hit an entity that is not the owner of this item
			if (!targetEntity) {
				//Box check doesn't care about non entities
				continue;
			}
			if (targetEntity === this.entity) {
				//Hit Self
				continue;
			}
			if (ignoreEntityIds.includes(targetEntity.id)) {
				continue;
			}
			this.Log("Hit Entity: " + targetEntity.id);

			//Raycast to the target to find a more concrete collisions\
			let rayStart = this.entity.GetHeadPosition();
			let rayEnd = targetEntity.GetHeadPosition();
			let hitDirection = rayEnd.sub(rayStart).normalized;

			//RAYCAST ALL
			this.Log("Raycast All");
			const hitInfosArray = Physics.RaycastAll(rayStart, hitDirection, rayDistance, -1);
			const hitInfos = hitInfosArray as unknown as CSArray<RaycastHit>;
			let blockerDistance = 9999;

			//Check each ray collision
			for (let i = 0; i < hitInfos.Length; i++) {
				this.Log("Raycasting to target");
				let hitInfo = hitInfos.GetValue(i);
				//Look for entities and blocking colliders
				const rayTarget = Entity.FindByCollider(hitInfo.collider);
				if (rayTarget) {
					if (rayTarget.id === this.entity.id) {
						//Hit self, skip
						continue;
					} else if (rayTarget.id === targetEntity.id) {
						this.Log("Raycast hit: " + rayTarget.id);
						//Raycast hit the target entity
						foundRaycastCollision = {
							hitEntity: targetEntity,
							hitDirection: hitDirection,
							hitPosition: hitInfo.point,
							hitNormal: hitInfo.normal,
							distance: hitInfo.distance,
							knockbackDirection: this.entity.gameObject.transform.forward,
						};
						// DebugUtil.DrawSingleLine(rayStart, hitInfo.point, Color.green, 2);
						if (!closestCollisionData || hitInfo.distance < closestCollisionData.distance) {
							this.Log("New closest target: " + hitInfo.distance);
							closestCollisionData = foundRaycastCollision;
							continue;
						}
					} else {
						//Hit a non entity object
						this.Log("Blocked by object: " + hitInfo.collider.gameObject.name);
						blockerDistance = math.min(blockerDistance, hitInfo.distance);
						// DebugUtil.DrawSingleLine(rayStart, hitInfo.point, Color.red, 2);
					}
				}

				if (foundRaycastCollision) {
					this.Log("Found target");
					if (foundRaycastCollision.distance < blockerDistance) {
						this.Log("Using target");
						//Not blocked by something
						collisionData.push(foundRaycastCollision);
						// DebugUtil.DrawSingleLine(rayStart, foundRaycastCollision.hitPosition, Color.green, 2);
					} else {
						this.Log("Can't use target because of blocker");
					}
				}
			}
		}

		if (this.meta.melee?.canHitMultipleTargets) {
			return collisionData;
		} else if (closestCollisionData) {
			return [closestCollisionData];
		}
		return [];
	}
}

export interface MeleeHit {
	hitEntity: Entity;
	hitDirection: Vector3;
	hitPosition: Vector3;
	hitNormal: Vector3;
	distance: number;
	knockbackDirection: Vector3;
}
